modules/nlu/src/backend/election/spellcheck-handler.ts

```diff
@@ -1,50 +1,41 @@
 
-import { BpPredictOutput as PredictOutput } from '../stan/api-mapper'
-import { NONE_INTENT, ValueOf } from './typings'
+const debug = DEBUG('nlu:predict')
 
-type Predictions = Dic<NLU.ContextPrediction>
-
-const mergeSpellChecked = (originalOutput: PredictOutput, spellCheckedOutput: PredictOutput): PredictOutput => {
-  const mostConfidentContext = (preds: Predictions): ValueOf<Predictions> =>
-    _(preds)
-      .values()
-      .maxBy(p => p.confidence)!
-
-  const mostConfidentIntent = (preds: ValueOf<Predictions>) =>
-    _(preds.intents)
-      .filter(i => i.label !== NONE_INTENT)
-      .maxBy(i => i.confidence)!
-
-  const originalPredictions: Predictions = originalOutput.predictions!
-  const spellCheckedPredictions: Predictions = spellCheckedOutput.predictions!
+const getTop1Intent = (mostConfidentCtx: NLU.ContextPrediction & { name: string }) => {
+  return _(mostConfidentCtx.intents)
+    .orderBy(i => i.confidence, 'desc')
+    .filter(i => i.label !== NONE_INTENT) // only relevant when legacy-election is enabled
+    .map(({ label, confidence }) => ({ name: label, context: mostConfidentCtx.name, confidence }))
+    .first()
+}
 
-  const mergeContextConfidence =
-    mostConfidentContext(originalPredictions).confidence < mostConfidentContext(spellCheckedPredictions).confidence
+const pickSpellChecked = (
+  originalOutput: IO.EventUnderstanding,
+  spellCheckedOutput: IO.EventUnderstanding
+): IO.EventUnderstanding => {
+  if (!originalOutput.predictions || !spellCheckedOutput.predictions) {
+    return originalOutput
+  }
 
-  for (const ctx of Object.keys(originalPredictions)) {
-    const originalPred = originalPredictions[ctx]
-    const spellCheckedPred = spellCheckedPredictions[ctx]
+  const originalMostConfidentContext = getMostConfidentContext(originalOutput)
+  const spellCheckedMostConfidentContext = getMostConfidentContext(spellCheckedOutput)
 
-    if (mergeContextConfidence) {
-      originalPred.confidence = _.mean([originalPred.confidence, spellCheckedPred.confidence])
-    }
+  if (!originalMostConfidentContext || !spellCheckedMostConfidentContext) {
+    return originalOutput
+  }
 
-    if (
-      originalPred.intents.length &&
-      mostConfidentIntent(originalPred).confidence <= mostConfidentIntent(spellCheckedPred).confidence
-    ) {
-      for (const intent of originalPred.intents) {
-        const originalIntent = originalPred.intents.find(i => i.label === intent.label)!
-        const spellCheckedIntent = spellCheckedPred.intents.find(i => i.label === intent.label)!
-        intent.confidence = _.max([originalIntent.confidence, spellCheckedIntent.confidence])!
-      }
-    }
+  const originalTop1 = getTop1Intent(originalMostConfidentContext)
+  const spellCheckedTop1 = getTop1Intent(spellCheckedMostConfidentContext)
 
-    originalPred.oos = spellCheckedPred.oos
+  if (!originalTop1 || !spellCheckedTop1 || originalTop1.confidence >= spellCheckedTop1.confidence) {
```

Comment:
We could also lower the confidence of a prediction if a spell-check was needed using some string distance metric or something based on the amount of token changed, but because this kind of logic was absent before, I don't feel the need to add it.